function.h

```diff
@@ -0,0 +1,57 @@
+#define WaterCherenkovAng 0.719887812
+#define WaterRefractiveIndex 1.33
+
+//return true when the main track has entrance or exit radius larger than the cutoff, otherwise return false
+//EntranceCut and ExitCut are the cutoff radius of the entrance and exit point
+bool selection(const double Entrance[3], const double Exit[3], const double EntranceCut = 120, const double ExitCut = 145) {
+    if ((Entrance[0]*Entrance[0]+Entrance[1]*Entrance[1])>(EntranceCut*EntranceCut)) {return true;}
+    if ((Exit[0]*Exit[0]+Exit[0]*Exit[0])>(ExitCut*ExitCut)) {return true;}
+    return false;
+}
+
+//
+std::vector<std::vector<double>> TrackReconstruction(const double VerPos[3], const double VerDir[3], const double PMTPos[2014][3], const double PMTOri[2014][3]) {
+    std::vector<std::vector<double>> PMTReTrack;
+    double VerToPMTDir[3], PhoDir[3], VerToPMTDist, VerToPhoDist, PhoToPMTDist, theta, HitTime, CosInAngle;
+    for (int i=0; i<2014; i++) {
+        //VerToPMTDir[3] is a unit vector pointing from the given vertex to the PMT
+        VerToPMTDir[0] = PMTPos[i][0] - VerPos[0];
+        VerToPMTDir[1] = PMTPos[i][1] - VerPos[1];
+        VerToPMTDir[2] = PMTPos[i][2] - VerPos[2];
+        VerToPMTDist = sqrt(VerToPMTDir[0]*VerToPMTDir[0] + VerToPMTDir[1]*VerToPMTDir[1] + VerToPMTDir[2]*VerToPMTDir[2]);
+        VerToPMTDir[0] = VerToPMTDir[0]/VerToPMTDist;
+        VerToPMTDir[1] = VerToPMTDir[1]/VerToPMTDist;
+        VerToPMTDir[2] = VerToPMTDir[2]/VerToPMTDist;
+
+        //theta is the angle between the vertex direction vector and the VerToPMTDir[3] vector
+        theta = VerToPMTDir[0]*VerDir[0] + VerToPMTDir[1]*VerDir[1] + VerToPMTDir[2]*VerDir[2];     //This is the value of cos(theta)
+        //Reconstruction fail when cos(theta) > 1
+        if (fabs(theta)>1) {
```

Comment:
numerical precision problem? for costh>1, it's better to have costh=1 instead of -1

Reply:
you mean manually setting costheta=1 if the value is slightly larger than 1 due to numerical precision problem?

Reply:
yes it is